protocols/kad/src/behaviour.rs

```diff
@@ -1475,9 +1514,16 @@ where
                         key, records, quorum, cache_at
                     } = &mut query.inner.info {
                         if let Some(record) = record {
-                            records.push(record);
+                            records.push((Some(source.clone()), record));
+
                             if records.len() >= quorum.get() {
-                                query.finish()
+                                // Desired quorum reached. The query may finish. See
+                                // [`Query::may_finish`] for details.
+                                let peers = records.iter()
+                                    .filter_map(|(peer, _record)| peer.as_ref())
+                                    .cloned()
+                                    .collect::<Vec<_>>();
```

Comment:
```suggestion
```
I think we can just pass in an iterator.

Reply:
Without the `cloned().collect()` we would borrow `query` twice.

```
error[E0499]: cannot borrow `*query` as mutable more than once at a time
    --> protocols/kad/src/behaviour.rs:1519:33
     |
1510 |                     } = &mut query.inner.info {
     |                         --------------------- first mutable borrow occurs here
...
1519 |                                 query.may_finish(peers)
     |                                 ^^^^^            ----- first borrow later used here
     |                                 |
     |                                 second mutable borrow occurs here

error: aborting due to previous error
```

Am I missing something?

Reply:
But at least `collect()` shouldn't be needed, or not? I would expect `.cloned()` to end the borrow on `records` already.

Reply:
Well, I'm probably mistaken, since the iterator needs to hold on to the source until iteration actually starts. Thanks for trying though!